File: anqa/harness/cursor.py
"""Cursor disk adapter (``~/.cursor/projects/*/agent-transcripts``).

One jsonl file is one conversation. Catalog path is ``cursor:<session_id>``.
List metadata comes from ``~/.cursor/chats/*/<id>/meta.json``.
"""

from __future__ import annotations

import json
import tarfile
from collections.abc import Iterator, Sequence
from datetime import UTC, datetime
from pathlib import Path

from .. import event_types as et
from ..models import JsonObject, SessionMeta, ToolInputBag, TraceEvent, as_json_object
from ..session.tagged_blocks import operator_prompt_text
from .ref import SessionRef

CURSOR_HARNESS_ID = "cursor"
_CANCELLED = frozenset({"cancelled", "canceled", "aborted", "interrupted", "error"})


def default_store_root() -> Path:
    """Host Cursor config tree (resolved at call time)."""
    return Path.home() / ".cursor"


def _as_object(raw: object) -> JsonObject:
    if isinstance(raw, dict):
        return as_json_object(raw)
    return {}


def _iso_ms(raw: object) -> str:
    if isinstance(raw, bool):
        return ""
    if isinstance(raw, (int, float)) and raw > 0:
        val = float(raw)
        sec = val / 1000.0 if val > 1e12 else val
        return datetime.fromtimestamp(sec, tz=UTC).strftime("%Y-%m-%dT%H:%M:%SZ")
    return ""


def _file_stamp(path: Path) -> tuple[float, int, int, int]:
    try:
        st = Path(path).expanduser().stat()
    except OSError:
        return (0.0, 0, 0, 0)
    return (float(st.st_mtime), int(st.st_size), 0, 0)


def _iter_rows(path: Path) -> Iterator[JsonObject]:
    for line in path.read_text(encoding="utf-8").splitlines():
        if not line.strip():
            continue
        try:
            raw = json.loads(line)
        except json.JSONDecodeError:
            continue
        if isinstance(raw, dict):
            yield as_json_object(raw)


def _load_json(path: Path) -> JsonObject:
    try:
        raw = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {}
    return _as_object(raw)


def _is_transcript(path: Path) -> bool:
    if not path.is_file() or path.suffix != ".jsonl":
        return False
    return path.parent.parent.name == "agent-transcripts"


def _collect_jsonl(roots: Sequence[Path]) -> list[Path]:
    out: list[Path] = []
    for raw in roots:
        path = Path(raw).expanduser()
        if _is_transcript(path):
            out.append(path)
            continue
        if not path.is_dir():
            continue
        out.extend(sorted(path.rglob("agent-transcripts/*/*.jsonl")))
    return out


def _find_meta(root: Path, sid: str) -> JsonObject:
    chats = root / "chats"
    if chats.is_dir():
        for path in chats.glob(f"*/{sid}/meta.json"):
            return _load_json(path)
    if root.name == sid and (root / "meta.json").is_file():
        return _load_json(root / "meta.json")
    for path in root.rglob(f"{sid}/meta.json"):
        return _load_json(path)
    return {}


def _blocks_text(raw: object) -> str:
    if isinstance(raw, str):
        return raw
    if not isinstance(raw, list):
        return ""
    bits: list[str] = []
    for item in raw:
        if not isinstance(item, dict):
            continue
        if str(item.get("type") or "") != "text":
            continue
        text = str(item.get("text") or "").strip()
        if text:
            bits.append(text)
    return "\n".join(bits)


def _session_title(header: JsonObject, rows: Sequence[JsonObject]) -> str:
    prompt = _first_user_title(rows)
    if prompt:
        return prompt
    raw = operator_prompt_text(str(header.get("title") or "").strip())
    return raw.splitlines()[0][:120] if raw else ""


def _first_user_title(rows: Sequence[JsonObject]) -> str:
    for row in rows:
        if str(row.get("role") or "") != "user":
            continue
        text = operator_prompt_text(_blocks_text(_as_object(row.get("message")).get("content")))
        if text:
            return text.splitlines()[0][:120]
    return ""


def _count_tools(rows: Sequence[JsonObject]) -> int:
    n = 0
    for row in rows:
        content = _as_object(row.get("message")).get("content")
        if not isinstance(content, list):
            continue
        for part in content:
            if isinstance(part, dict) and str(part.get("type") or "") == "tool_use":
                n += 1
    return n


def _last_signal(rows: Sequence[JsonObject]) -> tuple[str, str]:
    last_type = ""
    last_status = ""
    last_role = ""
    for row in rows:
        typ = str(row.get("type") or "").strip()
        if typ == "turn_ended":
            last_type = typ
            last_status = str(row.get("status") or "").strip()
            last_role = ""
            continue
        role = str(row.get("role") or "").strip()
        if role in {"user", "assistant"}:
            last_role = role
            last_type = ""
            last_status = ""
    return last_type, last_status or last_role


def _turn_outcome(rows: Sequence[JsonObject]) -> str:
    kind, status = _last_signal(rows)
    if kind == "turn_ended":
        if status.casefold() in _CANCELLED:
            return "cancelled"
        return "complete"
    if status in {"user", "assistant"}:
        return "running"
    return ""


def _from_message(index: int, row: JsonObject) -> list[TraceEvent]:
    role = str(row.get("role") or "")
    content = _as_object(row.get("message")).get("content")
    events: list[TraceEvent] = []
    if role == "user":
        text = operator_prompt_text(_blocks_text(content))
        events.append(
            TraceEvent(
                index=index,
                event_type=et.USER_MESSAGE_CHUNK,
                content=text,
                update_index=index,
            )
        )
        return events
    if role != "assistant":
        return events
    text = _blocks_text(content)
    if text:
        events.append(
            TraceEvent(
                index=index,
                event_type=et.AGENT_MESSAGE_CHUNK,
                content=text,
                update_index=index,
            )
        )
    if not isinstance(content, list):
        return events
    for part in content:
        if not isinstance(part, dict):
            continue
        if str(part.get("type") or "") != "tool_use":
            continue
        name = str(part.get("name") or "").strip()
        raw = part.get("input")
        bag = _as_object(raw) if isinstance(raw, dict) else {}
        events.append(
            TraceEvent(
                index=index,
                event_type=et.TOOL_CALL,
                tool_name=name,
                tool_call_id=str(part.get("id") or ""),
                raw_input=ToolInputBag(bag),
                update_index=index,
            )
        )
    return events


def _from_row(index: int, row: JsonObject) -> list[TraceEvent]:
    typ = str(row.get("type") or "")
    if typ == "turn_ended":
        status = str(row.get("status") or "").strip()
        ended = status.casefold() in _CANCELLED
        return [
            TraceEvent(
                index=index,
                event_type=et.TURN_ENDED if ended else et.TURN_COMPLETED,
                content=status,
                update_index=index,
            )
        ]
    if str(row.get("role") or ""):
        return _from_message(index, row)
    return []


def _timeline_for(rows: Sequence[JsonObject]) -> list[TraceEvent]:
    events: list[TraceEvent] = []
    for i, row in enumerate(rows):
        events.extend(_from_row(i, row))
    return events


def _meta_from(rows: Sequence[JsonObject], path: Path, sid: str, header: JsonObject) -> SessionMeta:
    created = _iso_ms(header.get("createdAtMs"))
    updated = _iso_ms(header.get("updatedAtMs")) or created
    start = None
    end = None
    created_raw = header.get("createdAtMs")
    updated_raw = header.get("updatedAtMs")
    if isinstance(created_raw, (int, float)) and not isinstance(created_raw, bool):
        start = float(created_raw)
        if start > 1e12:
            start /= 1000.0
    if isinstance(updated_raw, (int, float)) and not isinstance(updated_raw, bool):
        end = float(updated_raw)
        if end > 1e12:
            end /= 1000.0
    duration = float(max(0, (end or 0) - (start or 0))) if start and end else 0.0
    return SessionMeta(
        session_id=sid,
        session_dir=path,
        model_id="unknown",
        title=_session_title(header, rows),
        created_at=created,
        updated_at=updated,
        duration_seconds=duration,
        run_dir=str(header.get("cwd") or "").strip(),
        num_events=len(_timeline_for(rows)),
        tool_call_count=_count_tools(rows),
        turn_outcome=_turn_outcome(rows),
        harness=CURSOR_HARNESS_ID,
    )


def _ref_for_file(path: Path, cwd: str = "") -> SessionRef | None:
    if not _is_transcript(path):
        return None
    sid = path.stem.strip()
    if not sid:
        return None
    return SessionRef(
        harness=CURSOR_HARNESS_ID,
        session_id=sid,
        locator=path,
        cwd=cwd,
    )


def _jsonl_from_ref(ref: SessionRef | Path | str, root: Path) -> tuple[Path, str]:
    if isinstance(ref, SessionRef):
        return Path(ref.locator), ref.session_id
    from .ref import parse_session_ref_string

    parsed = parse_session_ref_string(str(ref))
    if parsed is not None:
        found = _find_file(root, parsed[1])
        return (found or Path(), parsed[1])
    path = Path(str(ref)).expanduser()
    if _is_transcript(path):
        return path, path.stem
    return Path(), path.name


def _find_file(root: Path, session_id: str) -> Path | None:
    sid = (session_id or "").strip()
    if not sid:
        return None
    for path in _collect_jsonl([root]):
        if path.stem == sid:
            return path
    return None


class CursorAdapter:
    """Read-only Cursor agent-transcript adapter."""

    id = CURSOR_HARNESS_ID
    product = "Cursor"
    supported_version = "2026.08.25-3e8eec8"

    def root(self) -> Path:
        return default_store_root()

    def default_host_roots(self) -> list[Path]:
        path = self.root()
        return [path] if path.is_dir() else []

    def discover(self, roots: Sequence[Path | str] | None = None) -> list[SessionRef]:
        scan = [self.root()] if roots is None else [Path(r) for r in roots]
        found: list[SessionRef] = []
        seen: set[str] = set()
        for file in _collect_jsonl(scan):
            header = _find_meta(self.root() if roots is None else scan[0], file.stem)
            ref = _ref_for_file(file, cwd=str(header.get("cwd") or "").strip())
            if ref is None or ref.session_id in seen:
                continue
            seen.add(ref.session_id)
            found.append(ref)
        return found

    def looks_like(self, ref: SessionRef | Path | str) -> bool:
        if isinstance(ref, SessionRef):
            return ref.harness == CURSOR_HARNESS_ID
        from .ref import parse_session_ref_string

        parsed = parse_session_ref_string(str(ref))
        if parsed is not None:
            return parsed[0] == CURSOR_HARNESS_ID
        return _is_transcript(Path(str(ref)).expanduser())

    def bind_locator(self, locator: Path) -> SessionRef | None:
        path = Path(locator).expanduser()
        if not self.looks_like(path):
            return None
        return _ref_for_file(path)

    def load_meta(self, ref: SessionRef | Path | str) -> SessionMeta:
        path, sid = _jsonl_from_ref(ref, self.root())
        if not path.is_file():
            raise FileNotFoundError(f"cursor session not found: {sid}")
        rows = list(_iter_rows(path))
        header = _find_meta(self.root(), sid)
        return _meta_from(rows, path, sid, header)

    def parse_timeline(self, ref: SessionRef | Path | str) -> list[TraceEvent]:
        path, _sid = _jsonl_from_ref(ref, self.root())
        if not path.is_file():
            return []
        return _timeline_for(list(_iter_rows(path)))

    def ref_for_id(self, session_id: str) -> SessionRef | None:
        sid = (session_id or "").strip()
        if not sid:
            return None
        found = _find_file(self.root(), sid)
        if found is None:
            return None
        return _ref_for_file(found)

    def watch_hints(self) -> tuple[str, ...]:
        return (".jsonl", "meta.json")

    def write_archive(self, ref: SessionRef | Path | str, dest: Path) -> list[str]:
        path, sid = _jsonl_from_ref(ref, self.root())
        if not path.is_file():
            raise FileNotFoundError(f"cursor session not found: {sid}")
        dest = Path(dest)
        dest.parent.mkdir(parents=True, exist_ok=True)
        tmp = dest.with_name(dest.name + ".tmp")
        packed = False
        members = [f"{sid}/{path.name}"]
        extras: list[tuple[Path, str]] = []
        for meta in self.root().joinpath("chats").glob(f"*/{sid}/meta.json"):
            extras.append((meta, f"{sid}/meta.json"))
            break
        try:
            with tarfile.open(tmp, "w:gz") as tf:
                tf.add(path, arcname=members[0])
                for extra, name in extras:
                    tf.add(extra, arcname=name)
                    members.append(name)
            tmp.replace(dest)
            packed = True
        except (OSError, tarfile.TarError) as exc:
            raise RuntimeError(f"failed to pack session archive: {exc}") from exc
        finally:
            if not packed:
                tmp.unlink(missing_ok=True)
        return members

    def open_archive(self, src: Path, dest_root: Path) -> SessionRef:
        from .grok import open_bound_archive

        return open_bound_archive(src, dest_root, self.bind_locator, harness=self.id)

    def load_detail(self, ref: SessionRef | Path | str) -> SessionMeta:
        return self.load_meta(ref)

    def timeline_stamp(self, ref: SessionRef | Path | str) -> tuple[float, int, int, int]:
        path, _sid = _jsonl_from_ref(ref, self.root())
        return _file_stamp(path)

    def trace_mtime(self, ref: SessionRef | Path | str) -> float:
        return self.timeline_stamp(ref)[0]

    def updates_size(self, ref: SessionRef | Path | str) -> int:
        return int(self.timeline_stamp(ref)[1])

    def scheduler_state(self, state: JsonObject) -> JsonObject | None:
        return None

    def reported_completion_ids(self, state: JsonObject) -> set[str]:
        return set()

    def list_turn_outcome(self, ref: SessionRef | Path | str) -> str:
        try:
            return (self.load_meta(ref).turn_outcome or "").strip()
        except FileNotFoundError:
            return ""


__all__ = [
    "CURSOR_HARNESS_ID",
    "CursorAdapter",
    "default_store_root",
]

File: core/src/stores/cursor.rs
//! Cursor agent-transcript jsonl.

use crate::event::{Event, EventType, ListMeta, ListStatus, SessionLocator};
use crate::jsonl::{self, JsonlRow};
use crate::store::Store;
use crate::text;
use serde_json::Value;
use std::path::{Path, PathBuf};

pub struct Cursor;

fn is_transcript(path: &Path) -> bool {
    path.is_file()
        && path.extension().and_then(|s| s.to_str()) == Some("jsonl")
        && path
            .parent()
            .and_then(|p| p.parent())
            .and_then(|p| p.file_name())
            .and_then(|n| n.to_str())
            == Some("agent-transcripts")
}

fn collect(roots: &[PathBuf]) -> Vec<PathBuf> {
    let mut out = Vec::new();
    for raw in roots {
        if is_transcript(raw) {
            out.push(raw.clone());
            continue;
        }
        if raw.is_dir() {
            out.extend(
                crate::walk::find_files(raw, ".jsonl", "")
                    .into_iter()
                    .filter(|p| p.components().any(|c| c.as_os_str() == "agent-transcripts")),
            );
        }
    }
    out
}

fn blocks_text(content: &Value) -> String {
    text::text_of(content)
}

fn user_query(text_in: &str) -> String {
    if let Some(start) = text_in.find("<user_query>") {
        if let Some(end) = text_in[start..].find("</user_query>") {
            return text_in[start + 12..start + end].trim().to_string();
        }
    }
    text_in.to_string()
}

fn from_row(row: &JsonlRow) -> Vec<Event> {
    let typ = text::field_str(&row.value, "type");
    if typ == "turn_ended" {
        let status = text::field_str(&row.value, "status");
        let ended = matches!(
            status.to_ascii_lowercase().as_str(),
            "cancelled" | "canceled" | "aborted" | "interrupted" | "error"
        );
        return vec![Event::new(if ended {
            EventType::TurnEnded
        } else {
            EventType::TurnCompleted
        })
        .with_content(status)
        .with_raw(&row.raw)];
    }
    let role = text::field_str(&row.value, "role");
    if role.is_empty() {
        return Vec::new();
    }
    let content = row
        .value
        .pointer("/message/content")
        .cloned()
        .unwrap_or(Value::Null);
    let mut events = Vec::new();
    if role == "user" {
        events.push(
            Event::new(EventType::UserMessageChunk)
                .with_content(user_query(&blocks_text(&content)))
                .with_raw(&row.raw),
        );
        return events;
    }
    if role != "assistant" {
        return events;
    }
    let text_body = blocks_text(&content);
    if !text_body.is_empty() {
        events.push(
            Event::new(EventType::AgentMessageChunk)
                .with_content(text_body)
                .with_raw(&row.raw),
        );
    }
    if let Some(parts) = content.as_array() {
        for part in parts {
            if text::field_str(part, "type") != "tool_use" {
                continue;
            }
            let name = text::field_str(part, "name");
            let input = part.get("input").cloned().unwrap_or(Value::Null);
            let mut ev = Event::new(EventType::ToolCall)
                .with_raw(serde_json::to_string(&input).unwrap_or_default());
            ev.tool_name = name;
            ev.tool_call_id = text::field_str(part, "id");
            events.push(ev);
        }
    }
    events
}

impl Store for Cursor {
    fn id(&self) -> &'static str {
        "cursor"
    }

    fn discover(&self, roots: &[PathBuf]) -> Vec<SessionLocator> {
        collect(roots)
            .into_iter()
            .filter_map(|file| {
                let sid = file.file_stem()?.to_str()?.to_string();
                Some(SessionLocator {
                    harness: "cursor".into(),
                    session_id: sid,
                    locator: file,
                    cwd: String::new(),
                })
            })
            .collect()
    }

    fn records(
        &self,
        locator: &Path,
        session_id: &str,
    ) -> Result<Vec<crate::store::Record>, String> {
        crate::store::jsonl_records(locator, self.id(), session_id)
    }

    fn events(&self, records: &[crate::store::Record]) -> Vec<Event> {
        records.iter().flat_map(from_row).collect()
    }

    fn list_meta(&self, locator: &Path, session_id: &str) -> Result<ListMeta, String> {
        if !locator.is_file() {
            return Err(format!("cursor session not found: {session_id}"));
        }
        let rows = jsonl::window(locator);
        let header = find_meta(locator, session_id);
        Ok(meta_from_window(&rows, locator, session_id, &header))
    }
}

fn load_json(path: &Path) -> Value {
    std::fs::read_to_string(path)
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or(Value::Null)
}

fn sid_meta_in(chats: &Path, sid: &str) -> Option<Value> {
    let Ok(entries) = std::fs::read_dir(chats) else {
        return None;
    };
    for entry in entries.flatten() {
        let meta = entry.path().join(sid).join("meta.json");
        if meta.is_file() {
            return Some(load_json(&meta));
        }
    }
    None
}

fn find_meta(locator: &Path, sid: &str) -> Value {
    for ancestor in locator.ancestors() {
        if let Some(v) = sid_meta_in(&ancestor.join("chats"), sid) {
            return v;
        }
    }
    if let Ok(home) = std::env::var("HOME") {
        if let Some(v) = sid_meta_in(&PathBuf::from(home).join(".cursor").join("chats"), sid) {
            return v;
        }
    }
    Value::Null
}

fn model_name_in(blob: &str) -> Option<String> {
    let needle = "\"modelName\"";
    let mut rest = blob;
    let mut found = None;
    while let Some(i) = rest.find(needle) {
        let after = rest[i + needle.len()..].trim_start();
        if let Some(after) = after.strip_prefix(':') {
            let after = after.trim_start();
            if let Some(after) = after.strip_prefix('"') {
                if let Some(end) = after.find('"') {
                    let name = after[..end].trim();
                    if !name.is_empty() {
                        found = Some(name.to_string());
                    }
                }
            }
        }
        rest = &rest[i + needle.len()..];
    }
    found
}

fn model_from_store(locator: &Path, sid: &str) -> String {
    let mut roots = Vec::new();
    for ancestor in locator.ancestors() {
        roots.push(ancestor.join("chats"));
    }
    if let Ok(home) = std::env::var("HOME") {
        roots.push(PathBuf::from(home).join(".cursor").join("chats"));
    }
    let mut found = String::new();
    for chats in roots {
        let Ok(entries) = std::fs::read_dir(&chats) else {
            continue;
        };
        for entry in entries.flatten() {
            let db = entry.path().join(sid).join("store.db");
            if !db.is_file() {
                continue;
            }
            let Ok(con) = rusqlite::Connection::open_with_flags(
                &db,
                rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY,
            ) else {
                continue;
            };
            let Ok(mut stmt) = con.prepare("SELECT data FROM blobs") else {
                continue;
            };
            let Ok(rows) = stmt.query_map([], |row| row.get::<_, Vec<u8>>(0)) else {
                continue;
            };
            for row in rows.flatten() {
                let text = String::from_utf8_lossy(&row);
                if let Some(name) = model_name_in(&text) {
                    found = name;
                }
            }
        }
    }
    found
}

fn first_user_title(rows: &[JsonlRow]) -> String {
    for row in rows {
        if text::field_str(&row.value, "role") != "user" {
            continue;
        }
        let content = row
            .value
            .pointer("/message/content")
            .cloned()
            .unwrap_or(Value::Null);
        let text_body = user_query(&blocks_text(&content));
        if !text_body.is_empty() {
            return text::first_line(&text_body, 120);
        }
    }
    String::new()
}

fn session_title(header: &Value, rows: &[JsonlRow]) -> String {
    let prompt = first_user_title(rows);
    if !prompt.is_empty() {
        return prompt;
    }
    let raw = user_query(&text::field_str(header, "title"));
    text::first_line(&raw, 120)
}

fn count_tools(rows: &[JsonlRow]) -> u32 {
    let mut n = 0u32;
    for row in rows {
        let content = row
            .value
            .pointer("/message/content")
            .cloned()
            .unwrap_or(Value::Null);
        let Some(parts) = content.as_array() else {
            continue;
        };
        n += parts
            .iter()
            .filter(|p| text::field_str(p, "type") == "tool_use")
            .count() as u32;
    }
    n
}

fn last_signal(rows: &[JsonlRow]) -> (String, String) {
    let mut last_type = String::new();
    let mut last_status = String::new();
    for row in rows {
        let typ = text::field_str(&row.value, "type");
        if typ == "turn_ended" {
            last_type = typ;
            last_status = text::field_str(&row.value, "status");
            continue;
        }
        let role = text::field_str(&row.value, "role");
        if role == "user" || role == "assistant" {
            last_type.clear();
            last_status = role;
        }
    }
    (last_type, last_status)
}

fn turn_outcome(rows: &[JsonlRow]) -> String {
    let (kind, status) = last_signal(rows);
    if kind == "turn_ended" {
        let mapped = ListStatus::from_token(&status);
        if mapped == ListStatus::Idle {
            return ListStatus::Complete.as_str().into();
        }
        return mapped.as_str().into();
    }
    ListStatus::from_token(&status).as_str().into()
}

fn meta_from_window(rows: &[JsonlRow], path: &Path, sid: &str, header: &Value) -> ListMeta {
    let created = text::field_iso(header, "createdAtMs");
    let mut updated = text::field_iso(header, "updatedAtMs");
    if updated.is_empty() {
        updated = created.clone();
    }
    let model = model_from_store(path, sid);
    ListMeta {
        session_id: sid.to_string(),
        locator: path.to_path_buf(),
        model_id: if model.is_empty() {
            "unknown".into()
        } else {
            model
        },
        title: session_title(header, rows),
        created_at: created.clone(),
        updated_at: updated.clone(),
        duration_seconds: text::duration_secs(
            text::epoch_secs(&header.get("createdAtMs").cloned().unwrap_or(Value::Null)),
            text::epoch_secs(&header.get("updatedAtMs").cloned().unwrap_or(Value::Null)),
        ),
        tool_call_count: count_tools(rows),
        turn_outcome: turn_outcome(rows),
        harness: "cursor".into(),
        harness_version: String::new(),
        run_dir: text::field_str(header, "cwd"),
        num_events: 0,
        has_subagents: false,
        subagent_count: 0,
        context_tokens_used: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn list_meta_window_title_and_last_turn() {
        let path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join(
            "../tests/fixtures/harness/cursor/projects/tmp-cursor-probe/agent-transcripts/aaaaaaaa-1111-4111-8111-000000000001/aaaaaaaa-1111-4111-8111-000000000001.jsonl",
        );
        let meta = Cursor
            .list_meta(&path, "aaaaaaaa-1111-4111-8111-000000000001")
            .unwrap();
        assert_eq!(meta.title, "Reply with CURSOR_PROBE_OK");
        assert_eq!(meta.turn_outcome, "complete");
        assert_eq!(meta.run_dir, "/tmp/cursor-probe-ws");
    }
}
